cli: teach dump about views and references

madelynnblue opened the pull request:
When a view is encountered, skip the data dumping phase.

References now cause the dump order to be sorted in a way that tables
depended on are output first.

We can now also remove the primary index detection since we support
directly sorting by it.

Fixes #16396
Fixes #17037

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17581)
<!-- Reviewable:end -->

knz commented:
This is brilliant! :fireworks: :100: 

Thank yo so much for doing this.

Regarding the dependency sort:

say a table C depends on B and B depends on A

The "dependedOnBy" (from `crdb_internal.forward_dependencies`) only contains the information "C -> B" and "B -> A", not "C -> A". This means that the sort function doesn't know that C comes after A. Is that a problem?

(I think it depends on the sort algorithm. If you're not sure, you need to populate the dependedOnBy with the transitive closure of all the dependency relations.)

---

Reviewed 3 of 3 files at r1.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17581#-:-KrFwLsFvT0kg9skkAIV:b5z1fkf)*
<!-- Sent from Reviewable.io -->

knz commented:
This means by the way you need to add a test with 3 tables with this relationship, perhaps in different name order combinations.

mjibson commented:
Yup, you're totally correct. Working on transitive deps.

mjibson commented:
PTAL

mjibson commented:
ping @knz

knz commented:
This works but is it not a bit inefficient?
If you want the list of tables in topological order, the simplest solution is to perform a depth-first traversal of the table dependency graph:
```go
func collect(tid ID, seen map[ID]bool, collected *[]ID) {
   for _, dep := range byID[tid].dependson {
        // depth-first collection of dependencies
        collect(dep, seen, collected)
   }
   // has this table already been collected previously?
   if _, ok := seen[tid]; ok { return }
   // no: add it
   *collected = append(*collected, tid)
   seen[tid] = true
}

// then somewhere else

var collected []ID
seen := make(map[ID]bool)
for tid := range selectedtables {
   collect(tid, seen, &collected)
}

// now collected is in topological order!
```

---

Reviewed 2 of 2 files at r2.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17581#-:-KreWraIEI0BlcK82wzV:bxm380v)*
<!-- Sent from Reviewable.io -->

knz commented:
hmm of course you don't need to look at the same table two times, so this is better:

```go
func collect(tid ID, seen map[ID]bool, collected *[]ID) {
   // has this table already been collected previously?
   if _, ok := seen[tid]; ok { return }

   for _, dep := range byID[tid].dependson {
        // depth-first collection of dependencies
        collect(dep, seen, collected)
   }
   // done: add the new table
   *collected = append(*collected, tid)
   seen[tid] = true
}```

madelynnblue committed:
cli: teach dump about views and references

When a view is encountered, skip the data dumping phase.

References now cause the dump order to be sorted in a way that tables
depended on are output first.

We can now also remove the primary index detection since we support
directly sorting by it.

Fixes #16396
Fixes #17037

mjibson commented:
Ah. Thanks, done.

---

Review status: 1 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17581#-:-KrgYMtpeo5NhzvAXq9o:b2ref3l)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo minor optimization

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/cli/dump.go, line 127 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17581#-KriYbTVi0QCGAAIvt-Y:-KriYbTWGVKuyfC3pW02:b4yuvl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc69ae0b2553bf3816755df9f77a93341c87f6f4/pkg/cli/dump.go#L127)):*
> ```Go
> 	}
> 	// has this table already been collected previously?
> 	if seen[tid] {
> ```

This check above the loop should be also OK and simplify slightly (to avoid traversing the deps of already-seen tables)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17581#-:-KriYhqDN3ZiWoRuy3NH:b-akb165)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/cli/dump.go, line 127 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17581#-KriYbTVi0QCGAAIvt-Y:-Krj2MOI7rMWrsxdVyN2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc69ae0b2553bf3816755df9f77a93341c87f6f4/pkg/cli/dump.go#L127)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This check above the loop should be also OK and simplify slightly (to avoid traversing the deps of already-seen tables)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17581)*
<!-- Sent from Reviewable.io -->

knz commented:
Either I made a mistake or the test was wrong :)

mjibson commented:
Tests pass locally so I have no idea what's going on. Retrying.

mjibson commented:
Problem was the dependency list was a map, which range iterates over in a non-deterministic order. Changed to a slice. Will merge when green.

knz commented:
:+1: thanks!!!
